src/Action/ViewPostAction.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\NewsBundle\Action;
+
+use Sonata\NewsBundle\Model\BlogInterface;
+use Sonata\NewsBundle\Model\PostManagerInterface;
+use Sonata\SeoBundle\Seo\SeoPageInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+final class ViewPostAction extends Controller
+{
+    /**
+     * @var BlogInterface
+     */
+    private $blog;
+
+    /**
+     * @var PostManagerInterface
+     */
+    private $postManager;
+
+    /**
+     * @var SeoPageInterface|null
+     */
+    private $seoPage;
+
+    /**
+     * ViewPostAction constructor.
+     *
+     * @param BlogInterface        $blog
+     * @param PostManagerInterface $postManager
+     */
+    public function __construct(BlogInterface $blog, PostManagerInterface $postManager)
+    {
+        $this->blog = $blog;
+        $this->postManager = $postManager;
+    }
+
+    /**
+     * @param string $permalink
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return Response
+     */
+    public function __invoke($permalink)
+    {
+        $post = $this->postManager->findOneByPermalink($permalink, $this->blog);
+
+        if (!$post || !$post->isPublic()) {
+            throw new NotFoundHttpException('Unable to find the post');
+        }
+
+        if ($seoPage = $this->seoPage) {
+            $seoPage
+                ->setTitle($post->getTitle())
+                ->addMeta('name', 'description', $post->getAbstract())
+                ->addMeta('property', 'og:title', $post->getTitle())
+                ->addMeta('property', 'og:type', 'blog')
+                ->addMeta('property', 'og:url', $this->generateUrl('sonata_news_view', [
+                    'permalink' => $this->getBlog()->getPermalinkGenerator()->generate($post),
+                ], UrlGeneratorInterface::ABSOLUTE_URL))
+                ->addMeta('property', 'og:description', $post->getAbstract())
+            ;
+        }
+
+        return $this->render('@SonataNews/Post/view.html.twig', [
+            'post' => $post,
+            'form' => false,
+            'blog' => $this->getBlog(),
+        ]);
+    }
+
+    /**
+     * @param null|SeoPageInterface $seoPage
+     */
+    public function setSeoPage(SeoPageInterface $seoPage = null)
```

Comment:
Is this needed? (It will be replaced with Trait?)

Reply:
Yes, after merging https://github.com/sonata-project/SonataSeoBundle/pull/274